cmd/provider/get.go

```diff
@@ -62,41 +64,30 @@ func beforeGetAdvertisements(cctx *cli.Context) error {
 		}
 	}
 
-	addr, err := multiaddr.NewMultiaddr(pAddrInfo)
+	provClient, err = toProviderClient(pAddrInfo, topic)
+	return err
+}
+
+func toProviderClient(addrStr string, topic string) (internal.ProviderClient, error) {
+	addr, err := multiaddr.NewMultiaddr(addrStr)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	addrInfos, err := peer.AddrInfosFromP2pAddrs(addr)
 	if err != nil {
-		return err
+		return nil, err
 	}
-
 	addrInfo := addrInfos[0]
-
-	var isHttp bool
 	for _, p := range addrInfo.Addrs[0].Protocols() {
 		if p.Code == multiaddr.P_HTTP || p.Code == multiaddr.P_HTTPS {
-			isHttp = true
-			break
+			return internal.NewHttpProviderClient(addrInfo)
 		}
 	}
 
-	if isHttp {
-		provClient, err = internal.NewHttpProviderClient(addrInfo)
-		if err != nil {
-			return err
-		}
-	} else {
-		if topic == "" {
-			return cli.Exit("topic must be configured when graphsync endpoint is specified.", 1)
-		}
-		provClient, err = internal.NewGraphSyncProviderClient(addrInfo, topic)
-		if err != nil {
-			return err
-		}
+	if topic == "" {
+		return nil, errors.New("topic must be configured when graphsync endpoint is specified")
```

Comment:
why is this? i thought topic was just for the gossiping - is it also used as the name of the `head` service?

Reply:
Yes for the head.

Reply:
there is i believe a way to enumerate supported protocols with multistream. i wonder if there's a sane default heuristic that we can use for guessing which one it probably is?

Reply:
My suggestion would be to update that in go-legs and reflect here. Cause right now creating a new syncer requires topic: https://github.com/filecoin-project/go-legs/blob/main/dtsync/sync.go#L103

Which I think is used only for head query.

Reply:
fair enough - though i guess the use in diagnostics is different from the main protocol where it should be known in advance